### PR TITLE
Broker log recovery

### PR DESCRIPTION
**Pull request by benbjohnson**
## Overview

This pull request adds recovery to the messaging.Topic when opening. If any partial messages are found then the file is truncated at that point and started from there. This can occur when ungracefully shutting down a server. It can leave half written messages at the end of segments.

### CODE REVIEW

File: messaging/broker.go [hunk 1]
```diff
@@ -1017,6 +1009,19 @@ func NewTopicReader(path string, index uint64, streaming bool) *TopicReader {
 	}
 }
 
+// Seek seeks to a position the current segment.
+func (r *TopicReader) Seek(offset int64, whence int) (int64, error) {
+	assert(whence == os.SEEK_CUR, "topic reader can only seek to a relative position")
```
Comment (otoolep) [inline messaging/broker.go]: Seems like you could remove this assert by simply having this function ignore `whence`, and then calling `Seek()` with `os.SEEK_CUR`. Of course, you may wish to catch errors in the future.

File: messaging/broker.go [hunk 2]
```diff
@@ -1285,8 +1297,14 @@ func (dec *MessageDecoder) Decode(m *Message) error {
 	m.unmarshalHeader(b[:])
 
 	// Read data.
-	if _, err := io.ReadFull(dec.r, m.Data); err != nil {
-		return fmt.Errorf("read body: %s", err)
+	if n, err := io.ReadFull(dec.r, m.Data); err == io.EOF || err == io.ErrUnexpectedEOF {
+		if _, err := dec.r.Seek(-int64(MessageHeaderSize+n), os.SEEK_CUR); err != nil {
+			return fmt.Errorf("cannot unread header+data: n=%d, err=%s", n, err)
+		}
+		warnf("unexpected eof(1): len=%d, n=%d, err=%s", len(m.Data), n, err)
```
Comment (otoolep) [inline messaging/broker.go]: Do you want this output here?

File: messaging/broker.go [hunk 3]
```diff
@@ -1265,19 +1270,26 @@ func (m *Message) unmarshalHeader(b []byte) {
...
-	if _, err := io.ReadFull(dec.r, b[:]); err == io.EOF {
+	// Unread if there is a partial read.
+	var b [MessageHeaderSize]byte
+	if n, err := io.ReadFull(dec.r, b[:]); err == io.EOF {
+		return err
+	} else if err == io.ErrUnexpectedEOF {
+		if _, err := dec.r.Seek(-int64(n), os.SEEK_CUR); err != nil {
+			return fmt.Errorf("cannot unread header: n=%d, err=%s", n, err)
+		}
+		warnf("unexpected eof(0): len=%d, n=%d, err=%s", len(b), n, err)
```
Comment (otoolep) [inline messaging/broker.go]: Ditto? Want this, or is it debug?

**Comment by otoolep** [comment]
+1, change looks good to me. I do have some questions about the `warnf` statements in there, whether you meant to keep them.

Don't forget the CHANGELOG.

**Comment by toddboom** [comment]
:+1:

File: messaging/broker.go (hunk 1 above)
Comment (benbjohnson) [inline messaging/broker.go]: If someone passes a non-`SEEK_CUR` in the future I want to catch it immediately. If I silently override it then it'll be a pain to debug.

File: messaging/broker.go (hunk 2 above)
Comment (benbjohnson) [inline messaging/broker.go]: Yeah, for the short term. It'll give us a little extra data. We don't have these variables returned to the caller and there's no logger in here so I couldn't think of a better way to inject it. We can remove the warns in the future.

**Pull request merged**